lib/elixir/lib/kernel/typespec.ex

```diff
@@ -248,8 +248,10 @@ defmodule Kernel.Typespec do
     end)
   end
 
-  defp translate_type({_, {kind, {:::, _, [{name, _, args}, definition]}, pos}}, state) do
+  defp translate_type({_, {kind, {:::, _, [{name, _, args}, definition]}, pos}}, global_state) do
     caller = :elixir_locals.get_cached_env(pos)
+    local_state = %{used_vars: []}
+    state = {global_state, local_state}
```

Comment:
My suggestion is, instead of a tuple, add a field called `local_vars: []` to the state which you always set to an empty list at the beginning. This way you don't need to pass a tuple around and do the packing/unpacking. Thoughts?

Reply:
The idea was to reveal intent, and keep local information locally. Otherwise, the state passed around in `translate_typespecs_for_module` would contain local information. We might empty the list before returning the results at the end, but I think it would be more obvious to do it at the beginning. 

Said that, your suggestion also sounds good because it simplifies the code and, as you said, the packing/unpacking.

Reply:
Yeah, let’s go with a single map because that’s also what we do for macro.env. :)